src/Storages/ExternalStream/Kafka/Kafka.cpp

```diff
@@ -276,45 +363,185 @@ void Kafka::validateMessageKey(const String & message_key_, IStorage * storage,
 }
 
 /// Validate the topic still exists, specified partitions are still valid etc
-void Kafka::validate(const std::vector<int32_t> & shards_to_query)
+void Kafka::validate()
 {
-    if (shards == 0)
-    {
-        std::scoped_lock lock(shards_mutex);
-        /// Recheck again in case in-parallel query chimes in and init the shards already
-        if (shards == 0)
-        {
-            /// We haven't describe the topic yet
-            auto result = getConsumer()->describe(settings->topic.value);
-            if (result.err != ErrorCodes::OK)
-                throw Exception(ErrorCodes::RESOURCE_NOT_FOUND, "{} topic doesn't exist", settings->topic.value);
+    auto consumer = getConsumer();
+    RdKafka::Topic rkt {*consumer->getHandle(), topic()};
+    auto parition_count = rkt.describe();
+    if (parition_count < 1)
+        throw Exception(ErrorCodes::INVALID_SETTING_VALUE, "Topic has no paritions, topic={}", topic());
 
-            shards = result.partitions;
-        }
-    }
-
-    if (!shards_to_query.empty())
+    if (!shards_from_settings.empty())
     {
         /// User specified specific partitions to consume.
         /// Make sure they are valid.
-        for (auto shard : shards_to_query)
+        for (auto shard : shards_from_settings)
         {
-            if (shard >= shards)
+            if (shard >= parition_count)
                 throw Exception(
                     ErrorCodes::INVALID_SETTING_VALUE,
-                    "Invalid topic partition {} for topic {}, biggest partition is {}",
+                    "Invalid topic partition {} for topic {}, total number of partitions is {}",
                     shard,
                     settings->topic.value,
-                    shards);
+                    parition_count);
         }
     }
 }
 
+Pipe Kafka::read(
+    const Names & column_names,
+    const StorageSnapshotPtr & storage_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr context,
+    QueryProcessingStage::Enum /*processed_stage*/,
+    size_t max_block_size,
+    size_t /*num_streams*/)
+{
+    auto consumer = getConsumer();
+    /// The topic_ptr can be shared between all the sources in the same pipe, because each source reads from a different partition.
+    auto topic_ptr = std::make_shared<RdKafka::Topic>(*consumer->getHandle(), topic());
+
+    /// User can explicitly consume specific kafka partitions by specifying `shards=` setting
+    /// `SELECT * FROM kafka_stream SETTINGS shards=0,3`
+    std::vector<int32_t> shards_to_query;
+    if (!shards_from_settings.empty())
+        shards_to_query = shards_from_settings;
+    else
+    {
+        /// The number of paritions could have been changed, always query the number before
+        auto parition_count = topic_ptr->describe();
```

Comment:
partition_count should be constants value, can we cache it ?

Reply:
Please the check the comment above. `paritition_count` does not change often for sure, but it's not a constant. Check #291